Split main package between the part that needs the solver and the part that does not

mmesiti opened the pull request:
With this work I wanted to check the possibility of separating the package in two parts, namely the one that needs absolutely OrdinaryDiffEq and its algorithms, and the one that does not, and create an umbrella package that contains both.

It is clear to me that from the point of view of the application domain of this code it does make little sense to have such a split, but from the development side it does, as the integration methods are actually outside of the package. 

If one just wants to split the code it is very easy - the only coupling to OrdinaryDiffEq happens when the integration method is passed as an argument.
But some of the current tests also require a proper integration method to be passed.

The work done here is the following:
1. refactored the `launchPMFRG!` function, to better show the data dependencies inside it, which can be then made more explicit with smaller functions;
2. Remove OrdinaryDiffEq as a dependency from the package, now renamed PMFRGCore
3. Create a new package, PMFRGSolve, which has instead OrdinaryDiffEq as a dependency and imports it.
4. Move tests and code from PMFRGCore to PMFRGSolve iteratively, following these steps:
    a. Run the tests for PMFRGCore
    b. take one/some tests that fail and the logic needed to make them pass into PMFRGSolve
    c. verify that now these tests as part of the test suite for PMFRGSolve do pass

Things to still do:
1. [X] verify that the total number of tests has not changed
2. [X] Move the example directory (which should be part of the umbrella package, not of any one below)
3. [ ] Try to run the examples and make sure they still work
4. [ ] This split has evidenced that most tests are of the end-to-end type, and that the PMFRGCore package in isolation is not really tested. It is ok but it should be easy to add some additional tests for, e.g., the calculation of the derivative (in many different cases: at the moment only the 1-loop case is tested).

Points for discussion:
- What if PMFRGSolve == PMFRG? i.e., moving the logic of PMFRGSolve in the umbrella package could make things simpler.
- I am aware that this might all be to an extent a futile exercise in refactoring, but after some tests are added to the PMFRGCore package, running the test suite for it could be much quicker (At the moment it's 7m41s for PMFRGSolve and 1m9s for PMFRGCore), and this should make development easier.

mmesiti committed:
use only a subset of OrdinaryDiffEq

This could speed (pre) compilation times
(and thus testing)
quite a lot? Hopefully?

mmesiti committed:
Split package: what needs DiffEq vs what doesnt

mmesiti committed:
WIP split packages + umbrella pkg, dev notes

mmesiti committed:
PMFRG -> PMFRGCore in the core package (to be checked again)

some tests for PMFRGCore still failing but expectedly
(they involve DP5(), which now is not visible inside PMFRGCore
because it does not import OrdinaryDiffEq)

mmesiti committed:
Splitting still WIP

need to decide how to split the solve method
between PMFRGCore and PMFRGSolve

mmesiti committed:
WIP refactoring solver

mmesiti committed:
formatting

mmesiti committed:
functions outside of launchpmfrg

mmesiti committed:
solver auxiliary funcs: Add types

to clarify the interface

mmesiti committed:
Fixed current PMFRGSolve tests

mmesiti committed:
moving failing tests from PMFRGCore to PMFRGSolve

failing tests being the ones that need OrdinaryDiffEq

mmesiti committed:
Separated test suites

not great, some coupling has becvome obvious

mmesiti committed:
Main package runs tests of subpackages

mmesiti committed:
Fixing github actions

The fact that the package needs SpinFRGLattices,
that is at the moment not registered,
and also some sub packages,
makes setting up the testing environment
a little tricky.

For example, when testing PMFRGSolve,
we need to have SpinFRGLattices available
(and we can Pkg.add it),
but when we try to add it
it will complain that PMFRGCore is not registered,
Adding PMFRGCore as a dev depenency
trigger the opposite problem:
an error message saying that SpinFRGLattices is not registered.

The solution is to add both packages as dev dependencies
at the same time,
using Pkg.develop
with a vector of PackageSpec objects.

mmesiti commented:
It turns out that there are at least a couple of negative aspects to this.

1. The build times for the PMFRGCore package, as they are now, are still about 7 minutes on GitHub actions. This might be due to the fact that even if the `OrdinaryDiffEq*` package containing the method is not needed, still other packages need to be loaded for some functionality in the current definition of PMFRGCore (probably they are actually not tested). 
In particular, the definition of the callbacks requires to add the DiffEqCallbacks package which has a number of dependencies, even if callbacks are never tested. 
This would require to move also the definition of the callbacks into the PMFRGSolve package, at the very least - it is not clear what the advantage would be. Perhaps disabling precompilation could reduce the time needed for testing, if some functionality is not actually used.

2. Testing of PMFRGSolve.jl fails saying that it expects PMFRGCore.jl to be registered. ATM I am not able to add a dev dependency to PMFRGSolve.jl before activating the environment, which seems to be a problem.
   _Edit: **the solution** to this is to add all the non-registered packages needed with a single `Pkg.develop()` call that takes as argument a vector of `PackageSpec` objects._
   In general, having to add the sub-packages as dev dependencies is a bit complicated though (perhaps there is a way to deal with this easily that I still have not seen).

mmesiti commented:
Update: Precompilation seems to happen twice, first ~when activating the environment~ when adding `SpinFRGLattices` as a dependency (from github) and then when testing. In both cases, it is reported that it takes around 170 seconds, and then testing takes around 70 seconds. Another job with precompilation reportedly takes 7m36s. 
Tried disabling precompilation, as previously seen elsewhere there is not a huge difference in time. The total time for setting up and test PMFRGCore is 6m45s (405 seconds), which is very similar to the time before when precompilation is active by default (~170 s* 2 + ~70 s = ~410s).

mmesiti commented:
It seem that disabling precompilation by default before launching Julia and then enabling just before calling Pkg.test() might be quicker. In this way, testing PMFRGCore takes 4m45s.

NilsNiggemann commented:
Aesthetically, I like the general Idea of separating the ODE solver from the core package, although it might be a bit of effort to disentangle everything.
Otherwise I am strongly in favor of naming the batteries-included main package which includes all dependencies "PMFRG.jl" ideally, so that from a user-perspective all the changes will not break anything.

I wonder if one can also use https://juliapackages.com/p/simplediffeq for speeding up tests, I guess that will no longer be necessary if we do have a separation.
of course it might cause some issues for development if a bug is introduced in PMFRGCore.jl which is not tested for and only throws an error in the ODE integration part. Probably that can be avoided well though.
I guess testing the derivative in the core package is then necessary to catch bugs as early as possible.

mmesiti commented:
> Aesthetically, I like the general Idea of separating the ODE solver from the core package, although it might be a bit of effort to disentangle everything.

Disentangling everything is more or less mindless work, once it is decided what needs to go in each package. There are some possible partially conflicting guiding principles to define what goes where:
- have as much logic as possible in the PMFRGCore package (the most conservative approach), making sure that as much code as possible is testable with unit tests and with minimal dependencies
- have all that remotely concerns the solver in PMFRGSolve, only the derivative calculations stay in PMFRGCore (by the way there's quite some logic in FileIO.jl that I am not sure where it would fit best)
- Reduce as much as possible the dependencies of PMFRGCore (so that no DiffEQ-related dependencies are added and the package can be tested in seconds)

Maybe it's impossible to define the best package subdivision, but we can get a first reasonable choice and then do some corrections if we think we need them.

> Otherwise I am strongly in favor of naming the batteries-included main package which includes all dependencies "PMFRG.jl" ideally, so that from a user-perspective all the changes will not break anything.

This was part of the original idea, just refactor the internals for development leaving an unchanged, clear interface for the user. In order for this to work without much effort from the user part, I think the sub-packages still need to be added to *a* registry as you already have done [here](https://github.com/NilsNiggemann/JuliaPMFRGRegistry). 

> I wonder if one can also use https://juliapackages.com/p/simplediffeq for speeding up tests, I guess that will no longer be necessary if we do have a separation. of course it might cause some issues for development if a bug is introduced in PMFRGCore.jl which is not tested for and only throws an error in the ODE integration part. Probably that can be avoided well though.

> I guess testing the derivative in the core package is then necessary to catch bugs as early as possible.

I agree that adding some tests in this regard should be paramount. My suggestion for that would be to add some characterization/regression tests by recording some input/output values during the tests (I still need to figure out the details on how to do this).

mmesiti committed:
WIP - Moving more dependencies to PMFRGSolve

(e.g. DiffEqCallbacks, SciMLBase)

mmesiti committed:
Remove unaccurate typing information

Actually in the case of Parquet,
SolveParquet does not pass a DiffEqCallbacks.SavedValues object
to saveMainOutput, but a simple StructArray.
So, requiring that saved_values always has a SavedValues type is incorrect.

mmesiti committed:
Formatting

mmesiti committed:
Adding a mechanism to regenerate expected values

The output values computed in the tests,
and compared to the expected results,
can be regenerated by setting the environment variable
"PMFRG_REGEN_EXPECTED_RESULTS" to any value
when running the tests.

mmesiti committed:
Adding a mechanism to regenerate expected values

The output values computed in the tests,
and compared to the expected results,
can be regenerated by setting the environment variable
"PMFRG_REGEN_EXPECTED_RESULTS" to any value
when running the tests.

mmesiti committed:
Merge branch 'exp-val-regen-tools' into core-solver-split

mmesiti committed:
testing IO only in PMFRGSolve

The saving functions depends on DiffEqCallbacks,
so they can only be tested there.

mmesiti committed:
Add comment in readme about regenerating exp vals

mmesiti committed:
Add mechanism to test readme for content

and some tests for readme

mmesiti committed:
Fixed generation of new expected results

- Missing `using Pkg` at the beginning now added
- Using type instead of value
  when creating new methods for example_Obs

mmesiti committed:
Removed compat constraints, updated expected values

mmesiti committed:
Merge branch 'exp-val-regen-tools' into core-solver-split

mmesiti committed:
Revert "testing IO only in PMFRGSolve"

This reverts commit 6d3961b45b523a407863c21309f94f982618d5e6.

This commit delete the wrong things

mmesiti committed:
testing IO only in PMFRGSolve

The saving functions depends on DiffEqCallbacks,
so they can only be tested there.
(this time hopefully correctly)

mmesiti committed:
Removing [compat] constraints

The compat constraints should have been removed
in the merge with exp-val-regen-tools,
but the merge was not done correctly.

mmesiti committed:
ExObsUpdate: proj/deps load conditional on version

The project/deps objects
that are created at the beginning
of the automatically generated ExampleObservable.jl file
belong to the Pkg API
which is not very stable,
so if they are read by different julia version
they might let to a crash of the test suite.

The representation of the proj/deps objects
has been sandwiched in an if statement
that allows the objects to be created
only if the julia VERSION
is the one used when generating the values.

Also, the code has been refactored in smaller functions.

mmesiti committed:
Merge branch 'exp-val-regen-tools' into core-solver-split

mmesiti committed:
Remove project/deps object, using Manifest instead

projec/deps objects not create any more in ExampleObservables.jl,
since the Pkg API is unstable.

Making a copy of the Manifest to be stored alongside
ExampleObservables.jl.

mmesiti committed:
ExampleObservable.jl: remove method redefs

Some methods were defined twice
because the regenerated file was not cleaned
before re-running the test suite.

Also, minor tweaks for readability
and to the documentation
(including tests)

mmesiti committed:
Merge branch 'exp-val-regen-tools' into core-solver-split

Also fixing some method redefinitions in the test suite
(changing their names between Solve and Core).
This error appears only when testing the whole PMFRG package.

mmesiti commented:
Note: this could be superseeded by #7 .
